File: scripts/ollama/pack_runtimes.py
#!/usr/bin/env python3
"""Pack pinned Ollama Linux runtimes from official archives."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import re
import shutil
import subprocess
import sys
import tarfile
from pathlib import Path
from typing import Dict, List, Optional, Sequence, Set

VERSION_RE = re.compile(r"^[0-9]+\.[0-9]+\.[0-9]+$")
SOURCE_FILES = {
    "linux-amd64": "ollama-linux-amd64.tar.zst",
    "linux-arm64": "ollama-linux-arm64.tar.zst",
}


SKIP_NAMES = {".DS_Store"}


def load_lock(root: Path) -> dict:
    lock_path = root / "ollama" / "runtimes.lock.json"
    with lock_path.open(encoding="utf-8") as handle:
        return json.load(handle)


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def file_size(path: Path) -> int:
    return path.stat().st_size


def member_matches(member: str, prefix: str) -> bool:
    return member == prefix or member.startswith(prefix.rstrip("/") + "/")


def list_members(archive: Path) -> List[str]:
    proc = subprocess.Popen(
        ["zstd", "-d", "-c", str(archive)],
        stdout=subprocess.PIPE,
    )
    assert proc.stdout is not None
    names: List[str] = []
    with tarfile.open(fileobj=proc.stdout, mode="r|") as tar:
        for item in tar:
            names.append(item.name)
    if proc.wait() != 0:
        raise SystemExit(f"failed to decompress {archive}")
    return names


def assert_layout(archive: Path, require: Sequence[str], forbid: Sequence[str]) -> None:
    names = list_members(archive)
    for needed in require:
        if not any(member_matches(name, needed) for name in names):
            raise SystemExit(f"{archive.name} is missing required path {needed}")
    for blocked in forbid:
        hits = [name for name in names if member_matches(name, blocked)]
        if hits:
            raise SystemExit(
                f"{archive.name} contains forbidden path {blocked}: {hits[0]}"
            )


def upstream_dir(root: Path) -> Path:
    return root / "ollama" / "upstream"


def download(url: str, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    partial = dest.with_name(dest.name + ".partial")
    if partial.exists():
        partial.unlink()
    subprocess.check_call(
        [
            "curl",
            "--fail",
            "--location",
            "--retry",
            "5",
            "--retry-delay",
            "2",
            "--connect-timeout",
            "20",
            "--output",
            str(partial),
            url,
        ]
    )
    partial.replace(dest)


def cache_matches(path: Path, sha256: str, size_bytes: int) -> bool:
    return (
        path.is_file()
        and file_size(path) == size_bytes
        and sha256_file(path) == sha256
    )


def ensure_source(source: dict, cache_dir: Path) -> Path:
    file_name = source["fileName"]
    dest = cache_dir / file_name
    if cache_matches(dest, source["sha256"], source["sizeBytes"]):
        print(f"using cached {file_name}", file=sys.stderr)
        return dest
    if dest.exists():
        print(f"replacing mismatched {file_name}", file=sys.stderr)
        dest.unlink()
    print(f"downloading {file_name}", file=sys.stderr)
    download(source["url"], dest)
    if not cache_matches(dest, source["sha256"], source["sizeBytes"]):
        dest.unlink(missing_ok=True)
        raise SystemExit(
            f"{file_name} did not match pinned sha256 or size after download"
        )
    return dest


def extract_excluding(archive: Path, dest: Path, exclude: Sequence[str]) -> None:
    if dest.exists():
        shutil.rmtree(dest)
    dest.mkdir(parents=True)
    command = ["tar", "--zstd", "-xf", str(archive), "-C", str(dest)]
    for prefix in exclude:
        command.extend(["--exclude", prefix])
    env = os.environ.copy()
    env["COPYFILE_DISABLE"] = "1"
    subprocess.check_call(command, env=env)
    for prefix in exclude:
        leftover = dest.joinpath(*prefix.split("/"))
        if leftover.exists():
            shutil.rmtree(leftover)


def should_skip(name: str) -> bool:
    base = Path(name).name
    return base in SKIP_NAMES or base.startswith("._")


def collect_tree(root: Path) -> List[Path]:
    entries: List[Path] = []
    for dirpath, dirnames, filenames in os.walk(root, followlinks=False):
        dirnames[:] = sorted(name for name in dirnames if not should_skip(name))
        filenames = sorted(name for name in filenames if not should_skip(name))
        current = Path(dirpath)
        if current != root:
            entries.append(current)
        for name in filenames:
            entries.append(current / name)
    entries.sort(key=lambda path: path.relative_to(root).as_posix())
    return entries


def write_tree_archive(tree: Path, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    if dest.exists():
        dest.unlink()
    proc = subprocess.Popen(
        ["zstd", "-q", "-T0", "-19", "-f", "-o", str(dest)],
        stdin=subprocess.PIPE,
    )
    assert proc.stdin is not None
    with tarfile.open(
        fileobj=proc.stdin, mode="w|", format=tarfile.GNU_FORMAT
    ) as tar:
        for path in collect_tree(tree):
            arcname = path.relative_to(tree).as_posix()
            info = tar.gettarinfo(str(path), arcname=arcname)
            info.uid = 0
            info.gid = 0
            info.uname = "root"
            info.gname = "root"
            info.mtime = 0
            if path.is_symlink():
                info.type = tarfile.SYMTYPE
                info.linkname = os.readlink(path)
                info.size = 0
                tar.addfile(info)
            elif path.is_dir():
                info.type = tarfile.DIRTYPE
                info.mode = 0o755
                info.size = 0
                tar.addfile(info)
            else:
                info.mode = 0o755 if os.access(path, os.X_OK) else 0o644
                with path.open("rb") as handle:
                    tar.addfile(info, handle)
    proc.stdin.close()
    if proc.wait() != 0:
        raise SystemExit(f"zstd failed writing {dest}")


def write_notices(root: Path, dest: Path) -> None:
    notices = dest / "OLLAMA_NOTICES.txt"
    parts = [
        (root / "ollama" / "NOTICES.md").read_text(encoding="utf-8"),
        "===== OLLAMA MIT LICENSE =====\n",
        (root / "LICENSES" / "ollama-MIT.txt").read_text(encoding="utf-8"),
    ]
    notices.write_text("\n".join(parts).rstrip() + "\n", encoding="utf-8")
    shutil.copy2(root / "LICENSES" / "ollama-MIT.txt", dest / "LICENSE")
    shutil.copy2(root / "ollama" / "runtimes.lock.json", dest / "OLLAMA_SOURCE.json")


def pack_one(
    pack: dict,
    source_path: Path,
    output: Path,
    work_dir: Path,
) -> Path:
    dest = output / pack["fileName"]
    transform = pack["transform"]
    if transform == "identity":
        if dest.exists() or dest.is_symlink():
            dest.unlink()
        try:
            os.link(source_path, dest)
        except OSError:
            shutil.copy2(source_path, dest)
    elif transform == "exclude":
        tree = work_dir / pack["id"]
        extract_excluding(source_path, tree, pack.get("exclude") or [])
        write_tree_archive(tree, dest)
        shutil.rmtree(tree)
    else:
        raise SystemExit(f"unsupported transform {transform} for {pack['id']}")
    assert_layout(dest, pack.get("require") or [], pack.get("forbid") or [])
    return dest


def selected_packs(lock: dict, pack_id: Optional[str]) -> List[dict]:
    packs = lock["packs"]
    if not pack_id:
        return packs
    matches = [pack for pack in packs if pack["id"] == pack_id]
    if not matches:
        raise SystemExit(f"unknown pack id {pack_id}")
    return matches


def parse_sha256sum(text: str) -> Dict[str, str]:
    mapping: Dict[str, str] = {}
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        parts = line.split()
        if len(parts) < 2:
            raise SystemExit(f"invalid sha256sum line: {line}")
        mapping[parts[-1].lstrip("*")] = parts[0].lower()
    return mapping


def curl_bytes(url: str) -> bytes:
    return subprocess.check_output(
        [
            "curl",
            "--fail",
            "--location",
            "--retry",
            "5",
            "--retry-delay",
            "2",
            "--connect-timeout",
            "20",
            url,
        ]
    )


def curl_content_length(url: str) -> int:
    headers = subprocess.check_output(
        ["curl", "--fail", "--location", "--silent", "--head", url],
        text=True,
    )
    length = None
    for line in headers.splitlines():
        key, _, value = line.partition(":")
        if key.lower() == "content-length" and value.strip().isdigit():
            length = int(value.strip())
    if not length:
        raise SystemExit(f"could not read Content-Length for {url}")
    return length


def official_url(version: str, file_name: str) -> str:
    return (
        f"https://github.com/ollama/ollama/releases/download/v{version}/{file_name}"
    )


def bump_runtimes(
    root: Path,
    version: str,
    license_file: Optional[Path] = None,
    checksums: Optional[Dict[str, str]] = None,
    sizes: Optional[Dict[str, int]] = None,
) -> None:
    if not VERSION_RE.match(version):
        raise SystemExit(f"version must be X.Y.Z, got {version}")
    lock = load_lock(root)
    lock["engineVersion"] = version
    lock["publicTag"] = f"ollama-v{version}"
    lock["displayName"] = f"Ollama Linux runtimes {version}"
    lock["upstreamTag"] = f"v{version}"
    lock["license"]["url"] = (
        f"https://raw.githubusercontent.com/ollama/ollama/v{version}/LICENSE"
    )

    if license_file is not None:
        license_bytes = license_file.read_bytes()
    else:
        license_bytes = curl_bytes(lock["license"]["url"])
    license_text = license_bytes.decode("utf-8")
    if not license_text.endswith("\n"):
        license_text += "\n"
        license_bytes = license_text.encode("utf-8")
    (root / "LICENSES" / "ollama-MIT.txt").write_bytes(license_bytes)
    lock["license"]["sha256"] = hashlib.sha256(license_bytes).hexdigest()

    if checksums is None:
        checksums = parse_sha256sum(
            curl_bytes(official_url(version, "sha256sum.txt")).decode("utf-8")
        )

    for source_id, file_name in SOURCE_FILES.items():
        source = lock["sources"][source_id]
        source["url"] = official_url(version, file_name)
        source["fileName"] = file_name
        if file_name not in checksums:
            raise SystemExit(f"{file_name} is not in upstream sha256sum.txt")
        source["sha256"] = checksums[file_name]
        if sizes is not None and file_name in sizes:
            source["sizeBytes"] = sizes[file_name]
        else:
            source["sizeBytes"] = curl_content_length(source["url"])

    lock_path = root / "ollama" / "runtimes.lock.json"
    lock_path.write_text(json.dumps(lock, indent=2) + "\n", encoding="utf-8")
    print(f"pinned Ollama {version} -> {lock_path}", file=sys.stderr)


def pack_runtimes(root: Path, output: Path, pack_id: Optional[str]) -> None:
    lock = load_lock(root)
    if output.exists():
        raise SystemExit(f"output path already exists: {output}")
    output.mkdir(parents=True)
    work_dir = output / ".work"
    work_dir.mkdir(parents=True)
    cache_dir = upstream_dir(root)
    cache_dir.mkdir(parents=True, exist_ok=True)

    packs = selected_packs(lock, pack_id)
    source_ids: Set[str] = {pack["source"] for pack in packs}
    sources: Dict[str, Path] = {}
    for source_id in sorted(source_ids):
        source = lock["sources"][source_id]
        sources[source_id] = ensure_source(source, cache_dir)

    for pack in packs:
        print(f"packing {pack['id']}", file=sys.stderr)
        pack_one(pack, sources[pack["source"]], output, work_dir)

    write_notices(root, output)
    shutil.rmtree(work_dir)
    checksum_lines = []
    for path in sorted(p for p in output.iterdir() if p.is_file()):
        checksum_lines.append(f"{sha256_file(path)}  {path.name}\n")
    (output / "SHA256SUMS").write_text("".join(checksum_lines), encoding="utf-8")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest="command", required=True)
    pack = sub.add_parser("pack", help="download official archives and emit runtime packs")
    pack.add_argument("--root", type=Path, required=True)
    pack.add_argument("--output", type=Path, required=True)
    pack.add_argument("--pack", dest="pack_id")
    bump = sub.add_parser("bump", help="retarget the lockfile to an official Ollama version")
    bump.add_argument("--root", type=Path, required=True)
    bump.add_argument("--version", required=True)
    bump.add_argument("--license-file", type=Path)
    return parser


def main(argv: Optional[Sequence[str]] = None) -> int:
    args = build_parser().parse_args(argv)
    if args.command == "pack":
        pack_runtimes(args.root.resolve(), args.output.resolve(), args.pack_id)
        return 0
    if args.command == "bump":
        license_file = args.license_file.resolve() if args.license_file else None
        bump_runtimes(args.root.resolve(), args.version, license_file)
        return 0
    raise SystemExit(f"unknown command {args.command}")


if __name__ == "__main__":
    sys.exit(main())

File: scripts/ollama/pack-runtimes.mjs
#!/usr/bin/env node
import { createHash } from "node:crypto";
import { spawn, spawnSync } from "node:child_process";
import { createReadStream, createWriteStream } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import { fileURLToPath } from "node:url";

const VERSION_RE = /^[0-9]+\.[0-9]+\.[0-9]+$/;
const SOURCE_FILES = {
  "linux-amd64": "ollama-linux-amd64.tar.zst",
  "linux-arm64": "ollama-linux-arm64.tar.zst",
  darwin: "ollama-darwin.tgz",
};
const SKIP_NAMES = new Set([".DS_Store"]);
const GNU_MAGIC = Buffer.from("ustar ");
const GNU_VERSION = Buffer.from(" \0");

export const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");

export function fail(message) {
  throw new Error(message);
}

export async function loadLock(root) {
  return JSON.parse(await fs.readFile(path.join(root, "ollama/runtimes.lock.json"), "utf8"));
}

export async function sha256File(file) {
  const hash = createHash("sha256");
  for await (const chunk of createReadStream(file)) {
    hash.update(chunk);
  }
  return hash.digest("hex");
}

export function memberMatches(member, prefix) {
  const trimmed = prefix.replace(/\/+$/, "");
  return member === trimmed || member.startsWith(`${trimmed}/`);
}

function toPosix(relativePath) {
  return relativePath.split(path.sep).join("/");
}

function shouldSkip(name) {
  return SKIP_NAMES.has(name) || name.startsWith("._");
}

function run(command, args, options = {}) {
  const result = spawnSync(command, args, {
    encoding: "utf8",
    ...options,
  });
  if (result.status !== 0) {
    const detail = (result.stderr || result.stdout || "").trim();
    fail(`${command} ${args.join(" ")} failed${detail ? `: ${detail}` : ""}`);
  }
  return result;
}

function tarListArgs(archive) {
  if (archive.endsWith(".tgz") || archive.endsWith(".tar.gz")) {
    return ["-tzf", archive];
  }
  if (archive.endsWith(".tar.zst")) {
    return ["--zstd", "-tf", archive];
  }
  fail(`unsupported archive ${archive}`);
}

export function listMembers(archive) {
  const result = run("tar", tarListArgs(archive), { stdio: ["ignore", "pipe", "pipe"] });
  return result.stdout
    .split(/\r?\n/)
    .filter(Boolean)
    .map((name) => name.replace(/^\.\//, ""));
}

export async function assertLayout(archive, require = [], forbid = []) {
  const names = listMembers(archive);
  for (const needed of require) {
    if (!names.some((name) => memberMatches(name, needed))) {
      fail(`${path.basename(archive)} is missing required path ${needed}`);
    }
  }
  for (const blocked of forbid) {
    const hit = names.find((name) => memberMatches(name, blocked));
    if (hit) {
      fail(`${path.basename(archive)} contains forbidden path ${blocked}: ${hit}`);
    }
  }
}

export function upstreamDir(root) {
  return path.join(root, "ollama/upstream");
}

async function sleep(ms) {
  await new Promise((resolve) => setTimeout(resolve, ms));
}

async function fetchBuffer(url) {
  let lastError;
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const response = await fetch(url, { redirect: "follow" });
      if (!response.ok) {
        throw new Error(`${url}: HTTP ${response.status}`);
      }
      return Buffer.from(await response.arrayBuffer());
    } catch (error) {
      lastError = error;
      await sleep(2000);
    }
  }
  throw lastError;
}

async function fetchContentLength(url) {
  const response = await fetch(url, { method: "HEAD", redirect: "follow" });
  if (!response.ok) {
    fail(`could not read Content-Length for ${url}`);
  }
  const length = Number(response.headers.get("content-length"));
  if (!Number.isInteger(length) || length <= 0) {
    fail(`could not read Content-Length for ${url}`);
  }
  return length;
}

async function download(url, dest) {
  await fs.mkdir(path.dirname(dest), { recursive: true });
  const partial = `${dest}.partial`;
  await fs.rm(partial, { force: true });
  let lastError;
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const response = await fetch(url, { redirect: "follow" });
      if (!response.ok) {
        throw new Error(`${url}: HTTP ${response.status}`);
      }
      if (!response.body) {
        throw new Error(`${url}: empty body`);
      }
      await pipeline(Readable.fromWeb(response.body), createWriteStream(partial));
      await fs.rename(partial, dest);
      return;
    } catch (error) {
      lastError = error;
      await fs.rm(partial, { force: true });
      await sleep(2000);
    }
  }
  throw lastError;
}

async function cacheMatches(file, sha256, sizeBytes) {
  try {
    const stat = await fs.stat(file);
    if (!stat.isFile() || stat.size !== sizeBytes) {
      return false;
    }
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }
    throw error;
  }
  return (await sha256File(file)) === sha256;
}

export async function ensureSource(source, cacheDir) {
  const fileName = source.fileName;
  const dest = path.join(cacheDir, fileName);
  if (await cacheMatches(dest, source.sha256, source.sizeBytes)) {
    console.error(`using cached ${fileName}`);
    return dest;
  }
  try {
    await fs.access(dest);
    console.error(`replacing mismatched ${fileName}`);
    await fs.unlink(dest);
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }
  console.error(`downloading ${fileName}`);
  await download(source.url, dest);
  if (!(await cacheMatches(dest, source.sha256, source.sizeBytes))) {
    await fs.rm(dest, { force: true });
    fail(`${fileName} did not match pinned sha256 or size after download`);
  }
  return dest;
}

async function extractExcluding(archive, dest, exclude) {
  await fs.rm(dest, { recursive: true, force: true });
  await fs.mkdir(dest, { recursive: true });
  const args = ["--zstd", "-xf", archive, "-C", dest];
  for (const prefix of exclude) {
    args.push("--exclude", prefix);
  }
  run("tar", args, {
    stdio: "inherit",
    env: { ...process.env, COPYFILE_DISABLE: "1" },
  });
  for (const prefix of exclude) {
    await fs.rm(path.join(dest, ...prefix.split("/")), { recursive: true, force: true });
  }
}

function writeString(buf, offset, max, value) {
  buf.fill(0, offset, offset + max);
  Buffer.from(String(value), "utf8").copy(buf, offset, 0, max);
}

function writeOctal(buf, offset, size, value) {
  buf.write(`${value.toString(8).padStart(size - 1, "0")}\0`, offset, size, "ascii");
}

function gnuHeader({ name, mode, size, typeflag, linkname = "" }) {
  const buf = Buffer.alloc(512);
  writeString(buf, 0, 100, name);
  writeOctal(buf, 100, 8, mode);
  writeOctal(buf, 108, 8, 0);
  writeOctal(buf, 116, 8, 0);
  writeOctal(buf, 124, 12, size);
  writeOctal(buf, 136, 12, 0);
  buf.fill(0x20, 148, 156);
  buf.write(typeflag, 156, 1, "ascii");
  writeString(buf, 157, 100, linkname);
  GNU_MAGIC.copy(buf, 257);
  GNU_VERSION.copy(buf, 263);
  writeString(buf, 265, 32, "root");
  writeString(buf, 297, 32, "root");
  let sum = 0;
  for (const byte of buf) {
    sum += byte;
  }
  buf.write(`${sum.toString(8).padStart(6, "0")}\0 `, 148, 8, "ascii");
  return buf;
}

function pad512(data) {
  const padded = Math.ceil(data.length / 512) * 512;
  if (padded === data.length) {
    return data;
  }
  return Buffer.concat([data, Buffer.alloc(padded - data.length)]);
}

function* gnuLongLink(name, typeflag) {
  const payload = Buffer.from(`${name}\0`, "utf8");
  yield gnuHeader({
    name: "././@LongLink",
    mode: 0o644,
    size: payload.length,
    typeflag,
  });
  yield pad512(payload);
}

async function collectTree(root) {
  const entries = [];
  async function walk(dir) {
    const names = (await fs.readdir(dir, { withFileTypes: true }))
      .filter((entry) => !shouldSkip(entry.name))
      .sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
    for (const entry of names) {
      const abs = path.join(dir, entry.name);
      entries.push(abs);
      const stat = await fs.lstat(abs);
      if (stat.isDirectory()) {
        await walk(abs);
      }
    }
  }
  await walk(root);
  entries.sort((a, b) => {
    const left = toPosix(path.relative(root, a));
    const right = toPosix(path.relative(root, b));
    return left < right ? -1 : left > right ? 1 : 0;
  });
  return entries;
}

async function* tarStream(tree) {
  for (const abs of await collectTree(tree)) {
    const rel = toPosix(path.relative(tree, abs));
    const stat = await fs.lstat(abs);
    if (rel.length >= 100) {
      yield* gnuLongLink(rel, "L");
    }
    if (stat.isSymbolicLink()) {
      const linkname = await fs.readlink(abs);
      if (linkname.length >= 100) {
        yield* gnuLongLink(linkname, "K");
      }
      yield gnuHeader({
        name: rel.slice(0, 100),
        mode: 0o777,
        size: 0,
        typeflag: "2",
        linkname: linkname.slice(0, 100),
      });
      continue;
    }
    if (stat.isDirectory()) {
      yield gnuHeader({
        name: rel.slice(0, 100),
        mode: 0o755,
        size: 0,
        typeflag: "5",
      });
      continue;
    }
    const mode = stat.mode & 0o111 ? 0o755 : 0o644;
    yield gnuHeader({
      name: rel.slice(0, 100),
      mode,
      size: stat.size,
      typeflag: "0",
    });
    const handle = await fs.open(abs, "r");
    try {
      const buf = Buffer.alloc(1024 * 1024);
      let remaining = stat.size;
      while (remaining > 0) {
        const { bytesRead } = await handle.read(buf, 0, Math.min(buf.length, remaining));
        if (bytesRead === 0) {
          fail(`short read while packing ${rel}`);
        }
        yield Buffer.from(buf.subarray(0, bytesRead));
        remaining -= bytesRead;
      }
    } finally {
      await handle.close();
    }
    const pad = (512 - (stat.size % 512)) % 512;
    if (pad) {
      yield Buffer.alloc(pad);
    }
  }
  yield Buffer.alloc(1024);
}

export async function writeTreeArchive(tree, dest) {
  await fs.mkdir(path.dirname(dest), { recursive: true });
  await fs.rm(dest, { force: true });
  const zstd = spawn("zstd", ["-q", "-T0", "-19", "-f", "-o", dest], {
    stdio: ["pipe", "inherit", "inherit"],
  });
  const closed = new Promise((resolve, reject) => {
    zstd.on("error", reject);
    zstd.on("close", (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(`zstd failed writing ${dest}`));
      }
    });
  });
  await pipeline(Readable.from(tarStream(tree)), zstd.stdin);
  await closed;
}

async function writeNotices(root, dest) {
  const notices = [
    await fs.readFile(path.join(root, "ollama/NOTICES.md"), "utf8"),
    "===== OLLAMA MIT LICENSE =====\n",
    await fs.readFile(path.join(root, "LICENSES/ollama-MIT.txt"), "utf8"),
  ];
  await fs.writeFile(path.join(dest, "OLLAMA_NOTICES.txt"), `${notices.join("\n").trimEnd()}\n`);
  await fs.copyFile(path.join(root, "LICENSES/ollama-MIT.txt"), path.join(dest, "LICENSE"));
  await fs.copyFile(
    path.join(root, "ollama/runtimes.lock.json"),
    path.join(dest, "OLLAMA_SOURCE.json"),
  );
}

async function packOne(pack, sourcePath, output, workDir) {
  const dest = path.join(output, pack.fileName);
  if (pack.transform === "identity") {
    await fs.rm(dest, { force: true });
    try {
      await fs.link(sourcePath, dest);
    } catch {
      await fs.copyFile(sourcePath, dest);
    }
  } else if (pack.transform === "exclude") {
    const tree = path.join(workDir, pack.id);
    await extractExcluding(sourcePath, tree, pack.exclude || []);
    await writeTreeArchive(tree, dest);
    await fs.rm(tree, { recursive: true, force: true });
  } else {
    fail(`unsupported transform ${pack.transform} for ${pack.id}`);
  }
  await assertLayout(dest, pack.require || [], pack.forbid || []);
  return dest;
}

function selectedPacks(lock, packId) {
  if (!packId) {
    return lock.packs;
  }
  const matches = lock.packs.filter((pack) => pack.id === packId);
  if (matches.length === 0) {
    fail(`unknown pack id ${packId}`);
  }
  return matches;
}

export function parseSha256Sum(text) {
  const mapping = {};
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#")) {
      continue;
    }
    const parts = line.split(/\s+/);
    if (parts.length < 2) {
      fail(`invalid sha256sum line: ${line}`);
    }
    mapping[parts[parts.length - 1].replace(/^\*/, "")] = parts[0].toLowerCase();
  }
  return mapping;
}

function officialUrl(version, fileName) {
  return `https://github.com/ollama/ollama/releases/download/v${version}/${fileName}`;
}

export async function bumpRuntimes(root, version, options = {}) {
  if (!VERSION_RE.test(version)) {
    fail(`version must be X.Y.Z, got ${version}`);
  }
  const lock = await loadLock(root);
  lock.engineVersion = version;
  lock.publicTag = `ollama-v${version}`;
  lock.displayName = `Ollama runtimes ${version}`;
  lock.upstreamTag = `v${version}`;
  lock.license.url = `https://raw.githubusercontent.com/ollama/ollama/v${version}/LICENSE`;

  let licenseBytes = options.licenseFile
    ? await fs.readFile(options.licenseFile)
    : await fetchBuffer(lock.license.url);
  let licenseText = licenseBytes.toString("utf8");
  if (!licenseText.endsWith("\n")) {
    licenseText += "\n";
    licenseBytes = Buffer.from(licenseText, "utf8");
  }
  await fs.writeFile(path.join(root, "LICENSES/ollama-MIT.txt"), licenseBytes);
  lock.license.sha256 = createHash("sha256").update(licenseBytes).digest("hex");

  const checksums =
    options.checksums ?? parseSha256Sum((await fetchBuffer(officialUrl(version, "sha256sum.txt"))).toString("utf8"));

  for (const [sourceId, fileName] of Object.entries(SOURCE_FILES)) {
    const source = lock.sources[sourceId];
    source.url = officialUrl(version, fileName);
    source.fileName = fileName;
    if (!(fileName in checksums)) {
      fail(`${fileName} is not in upstream sha256sum.txt`);
    }
    source.sha256 = checksums[fileName];
    source.sizeBytes =
      options.sizes && fileName in options.sizes
        ? options.sizes[fileName]
        : await fetchContentLength(source.url);
  }

  const lockPath = path.join(root, "ollama/runtimes.lock.json");
  await fs.writeFile(lockPath, `${JSON.stringify(lock, null, 2)}\n`);
  console.error(`pinned Ollama ${version} -> ${lockPath}`);
}

export async function packRuntimes(root, output, packId) {
  const lock = await loadLock(root);
  let outputExists = false;
  try {
    await fs.access(output);
    outputExists = true;
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }
  if (outputExists) {
    fail(`output path already exists: ${output}`);
  }
  await fs.mkdir(output, { recursive: true });
  const workDir = path.join(output, ".work");
  await fs.mkdir(workDir, { recursive: true });
  const cacheDir = upstreamDir(root);
  await fs.mkdir(cacheDir, { recursive: true });

  const packs = selectedPacks(lock, packId);
  const sourceIds = [...new Set(packs.map((pack) => pack.source))].sort();
  const sources = {};
  for (const sourceId of sourceIds) {
    sources[sourceId] = await ensureSource(lock.sources[sourceId], cacheDir);
  }
  for (const pack of packs) {
    console.error(`packing ${pack.id}`);
    await packOne(pack, sources[pack.source], output, workDir);
  }
  await writeNotices(root, output);
  await fs.rm(workDir, { recursive: true, force: true });

  const files = (await fs.readdir(output, { withFileTypes: true }))
    .filter((entry) => entry.isFile())
    .map((entry) => entry.name)
    .sort();
  const checksumLines = [];
  for (const name of files) {
    checksumLines.push(`${await sha256File(path.join(output, name))}  ${name}\n`);
  }
  await fs.writeFile(path.join(output, "SHA256SUMS"), checksumLines.join(""));
}

function parseArgs(argv) {
  const args = { _: [] };
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (token === "--root") {
      args.root = argv[(i += 1)];
    } else if (token === "--output") {
      args.output = argv[(i += 1)];
    } else if (token === "--pack") {
      args.pack = argv[(i += 1)];
    } else if (token === "--license-file") {
      args.licenseFile = argv[(i += 1)];
    } else if (token.startsWith("-")) {
      fail(`unknown flag ${token}`);
    } else {
      args._.push(token);
    }
  }
  return args;
}

async function main(argv = process.argv.slice(2)) {
  const args = parseArgs(argv);
  const command = args._.shift();
  const root = path.resolve(args.root || repoRoot);
  if (command === "pack") {
    const output = args.output || args._[0];
    if (!output) {
      fail("usage: pack-runtimes.mjs pack <output-dir> [pack-id]");
    }
    await packRuntimes(root, path.resolve(output), args.pack || args._[1]);
    return;
  }
  if (command === "bump") {
    const version = args._[0];
    if (!version) {
      fail("usage: pack-runtimes.mjs bump <x.y.z>");
    }
    await bumpRuntimes(root, version, {
      licenseFile: args.licenseFile ? path.resolve(args.licenseFile) : undefined,
    });
    return;
  }
  fail("usage: pack-runtimes.mjs <pack|bump> ...");
}

function invokedDirectly() {
  const entry = process.argv[1];
  if (!entry) {
    return false;
  }
  return path.resolve(entry) === fileURLToPath(import.meta.url);
}

if (invokedDirectly()) {
  main().catch((error) => {
    console.error(error.message);
    process.exit(1);
  });
}
